File: algorithm/22.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>

char **ans;
int left;
int right;
int rear;

void parenthesisUtil(int n, int *returnSize, int k, char *ele)
{
    if (left == n && right == n)
    {
        (*returnSize)++;
        ans[(*returnSize) - 1] = (char *)malloc(2 * n * sizeof(char));
        memcpy(ans[(*returnSize) - 1], ele, 2 * n * sizeof(char));
        return;
    }
    if (left < n)
    {
        ele[rear++] = '(';
        left++;
        parenthesisUtil(n, returnSize, k + 1, ele);
        rear--;
        left--;
    }
    if (left > right && right < n)
    {
        ele[rear++] = ')';
        right++;
        parenthesisUtil(n, returnSize, k + 1, ele);
        rear--;
        right--;
    }
}

char **generateParenthesis(int n, int *returnSize)
{
    char *ele = (char *)malloc(2 * n * sizeof(char));
    ans = (char **)malloc(2001 * sizeof(char *));
    left = 0;
    right = 0;
    rear = 0;
    (*returnSize) = 0;
    parenthesisUtil(n, returnSize, 0, ele);
    return ans;
}

int main(int argc, char *argv[])
{
    int n = 3;
    int *returnSize = (int *)calloc(1, sizeof(int));
    char **res = generateParenthesis(n, returnSize);
    for (int i = 0; i < (*returnSize); i++)
    {
        printf("%s\n", res[i]);
    }
    return 0;
}

File: go/goalgor/algor/22.go
package main

import (
	"fmt"
	"strings"
)

func generateParenthesis(n int) []string {
	left := make([]int, n)
	right := make([]int, n)
	results := []string{}
	ans := []string{}
	var iterate func()
	iterate = func() {
		if len(left) == 0 && len(right) == 0 {
			results = append(results, strings.Join(ans, ""))
			return
		}
		if len(left) != 0 && len(right) == 0 {
			return
		}
		if len(left) < len(right) {
			// Add left
			if len(left) != 0 {
				ans = append(ans, "(")
				left = left[:len(left)-1]
				iterate()
				ans = ans[:len(ans)-1]
				left = append(left, 0)
			}
			// Add Right
			ans = append(ans, ")")
			right = right[:len(right)-1]
			iterate()
			ans = ans[:len(ans)-1]
			right = append(right, 0)
		} else if len(left) == len(right) {
			// Add left
			ans = append(ans, "(")
			left = left[:len(left)-1]
			iterate()
			ans = ans[:len(ans)-1]
			left = append(left, 0)
		}
	}
	iterate()
	return results
}

func main() {
	fmt.Println(generateParenthesis(3))
}
